internal/controller/apiextensions/composite/api.go

```diff
@@ -113,6 +119,107 @@ func (a *APIFilteredSecretPublisher) UnpublishConnection(_ context.Context, _ re
+	return comp, errors.Wrap(err, errGetComposition)
+}
+
+// An APIRevisionFetcher selects the appropriate CompositionRevision for a
+// composite resource, fetches it, and returns it as a Composition. This is done
+// for compatibility with existing Composition logic while CompositionRevisions
+// are in alpha.
+type APIRevisionFetcher struct {
+	ca resource.ClientApplicator
+}
+
+// NewAPIRevisionFetcher returns a RevisionFetcher that fetches the
+// Revision referenced by a composite resource.
+func NewAPIRevisionFetcher(ca resource.ClientApplicator) *APIRevisionFetcher {
+	return &APIRevisionFetcher{ca: ca}
+}
+
+// Fetch the appropriate CompositionRevision for the supplied composite
+// resource and convert it to a Composition. Panics if the composite resource's
+// composition reference is nil, but handles setting the composition revision
+// reference.
+func (f *APIRevisionFetcher) Fetch(ctx context.Context, cr resource.Composite) (*v1.Composition, error) { //nolint:gocyclo
+	// NOTE(negz): This method is slightly over our complexity goal, but it
+	// is intended to be temporary and is readable as is. If it lives past
+	// the alpha stage of CompositionRevision support it may be worth
+	// breaking CompositionRevision selection out into a separate method.
+
+	ref := cr.GetCompositionRevisionReference()
+	pol := cr.GetCompositionUpdatePolicy()
+
+	// We've already selected a revision, and our update policy is manual.
+	// Just fetch and return the selected revision.
+	if ref != nil && pol != nil && *pol == xpv1.UpdateManual {
+		rev := &v1alpha1.CompositionRevision{}
+		err := f.ca.Get(ctx, meta.NamespacedNameOf(ref), rev)
+		return AsComposition(rev), errors.Wrap(err, errGetCompositionRevision)
+	}
+
+	// We either haven't yet selected a revision, or our update policy is
+	// automatic. Either way we need to determine the latest revision.
+
+	comp := &v1.Composition{}
+	if err := f.ca.Get(ctx, meta.NamespacedNameOf(cr.GetCompositionReference()), comp); err != nil {
+		return nil, errors.Wrap(err, errGetComposition)
+	}
+
+	rl := &v1alpha1.CompositionRevisionList{}
+	if err := f.ca.List(ctx, rl); err != nil {
+		return nil, errors.Wrap(err, errListCompositionRevisions)
+	}
+
+	currentHash := comp.Spec.Hash()
+
+	var latestIdx int = -1
+	for i := range rl.Items {
+		if !metav1.IsControlledBy(&rl.Items[i], comp) {
+			continue
+		}
+
+		// We use the hash, not the revision number, to determine which
+		// revision is current. This is because the Composition may have
+		// been reverted to an older state. When this happens we don't
+		// create a new CompositionRevision; rather the prior revision
+		// becomes effectively 'current'.
+		if rl.Items[i].GetLabels()[v1alpha1.LabelCompositionSpecHash] == currentHash {
+			latestIdx = i
+		}
+	}
+
+	if latestIdx == -1 {
+		return nil, errors.New(errNoCompatibleCompositionRevision)
+	}
+
+	latest := &rl.Items[latestIdx]
```

Comment:
After reading the code here, I'm wondering if there is any functionality of the `revision` field in the `CompositionRevision` spec other than just showing how many revisions were created (excepts reverts to an older state). 
I was expecting it to be somehow used to find the latest but at the same time, I was aware that wouldn't work in case of reverts. Now, I am wondering if we could drop that `revision` field completely or make it functional (i.e. rely on it to find the latest here) by also increasing it in case of reverts.